ml/Host.cc

```diff
@@ -376,11 +387,32 @@ void DetectableHost::detectOnce() {
             MaxTrainingDuration = std::max(MaxTrainingDuration, DimTrainingDuration);
             TotalTrainingDuration += DimTrainingDuration;
 
-            if (IsAnomalous)
+            if (IsAnomalous) {
                 NumAnomalousDimensions += 1;
+                /*count up the number of anomalies for this dimension*/
+                D->setAnomalousBitCount(D->getAnomalousBitCount() + 1);
+            }
+
+            /*regardless the dimension value was anomalous or not, update the value of the percentage of anomalous dimension*/            
+            D->setAnomalyPercentage((D->getAnomalousBitCount() / (abs(Cfg.SaveAnomalyPercentageEvery - AnomalyBitCounterWindow) * static_cast<double>(updateEvery()))) * 100.0);
+            /*Register the oldest time of this dimension*/
+            OldestTimeOfAllDims = MIN(D->oldestTime(), OldestTimeOfAllDims);
+            
+            /*if the counting window is exhausted, push and then reset the counter*/
+            if(AnomalyBitCounterWindow == 0) {
+                double AnomalyPercentage = (D->getAnomalousBitCount() / (Cfg.SaveAnomalyPercentageEvery * static_cast<double>(updateEvery()))) * 100.0;
+                DimsAnomalyRate.push_back({AnomalyPercentage , D->getID() });                
+                D->setAnomalousBitCount(0.0);
+
+                //Delete the old records based on the oldest time of dim data in dbengine, i.e. OldestTimeOfAllDims
+                DB.removeOldAnomalyRateInfo(OldestTimeOfAllDims);
+                info("ML Anomaly Rate Info: Deleted records older than %ld", OldestTimeOfAllDims);
```

Comment:
Deletion of data should be done outside of the locked region, ie. outside of the scope of `std::lock_guard<std::mutex> Lock(Mutex);`

Reply:
but that would implicate another additional loop over all dimensions!

Reply:
Why? At the end of this loop you'll have the oldest time in the variable `OldestTimeOfAllDims`. You can use it to call `DB.removeOldAnomalyRateInfo(OldestTimeOfAllDims)`

Reply:
OK, of course, done.